include/nlohmann/detail/output/serializer.hpp

```diff
@@ -811,6 +811,17 @@ class serializer
         return state;
     }
 
+    inline number_unsigned_t remove_sign(number_unsigned_t x) noexcept
+    {
+        return x;
+    }
+
+    inline number_unsigned_t remove_sign(number_integer_t x) noexcept
+    {
+        assert(x < std::numeric_limits<number_integer_t>::max());
```

Comment:
This assert doesn't make sense. Any value of type `number_integer_t` will be smaller than (or equal to) its max.

Even if it's a typo (and should be `number_unsigned_t`), I still am not sure what this is guarding. Are you trying to avoid that `sizeof(number_unsigned_t) < sizeof(number_integer_t)`?

Reply:
@jaredgrubb

I'm adding 1 to x in the line below, and that does only work if x < max.

Reply:
Right, but this function only gets called if the value is negative. I think you shouldn't have factored out this as a function because that fact gets lost.

Reply:
If I don't factor it out the compilers complain because I'm taking the negative of an unsigned value.

See

```
C:\projects\json\single_include\nlohmann/json.hpp(14268): warning C4146: unary minus operator applied to unsigned type, result still unsigned [C:\projects\json\test\test-algorithms.vcxproj]
  C:\projects\json\single_include\nlohmann/json.hpp(13895): note: see reference to function template instantiation 'void nlohmann::detail::serializer<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>::dump_integer<unsigned __int64,0>(NumberType)' being compiled
          with
          [
              NumberType=unsigned __int64
          ]

```

from https://ci.appveyor.com/project/nlohmann/json/builds/26925012/job/5jasuk4xo67e7o39.

I can add another assertion with `x < 0`;

Reply:
@jaredgrubb

I've rewritten the code a bit so that it is less confusing.